src/ray/raylet/cgroup_setup.cc

```diff
@@ -0,0 +1,80 @@
+// Copyright 2024 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "src/ray/raylet/cgroup_setup.h"
+
+#include <sys/stat.h>
+
+#include <cerrno>
+#include <cstring>
+#include <fstream>
+#include <string_view>
+#include <vector>
+
+#include "src/ray/util/logging.h"
+
+namespace ray {
+
+namespace {
+// Cgroup path for ray system components.
+constexpr std::string_view kSystemCgroupFolder = "/sys/fs/cgroup/system";
+// Cgroup PID path for ray system components.
+constexpr std::string_view kSystemCgroupProcs = "/sys/fs/cgroup/system/cgroup.procs";
+// Parent cgroup path.
+constexpr std::string_view kRtootCgroupProcs = "/sys/fs/cgroup/cgroup.procs";
+// Cgroup subtree control path.
+constexpr std::string_view kRootCgroupSubtreeControl =
+    "/sys/fs/cgroup/cgroup.subtree_control";
+// Owner can read and write.
+constexpr int kCgroupFilePerm = 0600;
+
+// If system processes are running in the container, all processes will be placed in the
+// root cgroup. This function will move all PIDs under root cgroup into system cgroup.
+void MoveProcsInSystemCgroup() {
```

Comment:
will making this function 'inline' help minimize performance penalty of this?

Reply:
I almost never use `inline` since compiler does better than humans.

inline helps reduce function invocation, mainly calling conventions;
but it also hurts i-cache locality and increases binary size.